File: main/binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        ret = []
        if root is None:
            return ret

        q = collections.deque((root,))
        while q:
            ret.append(q[-1].val)
            for _ in range(len(q)):
                root = q.popleft()
                if root.left is not None:
                    q.append(root.left)
                if root.right is not None:
                    q.append(root.right)

        return ret

File: main/binary-tree-right-side-view/binary-tree-right-side-view.java
import java.util.Deque;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        final List<Integer> ret = new ArrayList<>();
        if (root == null) return ret;

        final Deque<TreeNode> q = new ArrayDeque<>();
        for (q.add(root); !q.isEmpty(); ) {
            ret.add(q.peekLast().val);
            for (int width = q.size(); width != 0; --width) {
                root = q.pop();
                if (root.left != null) q.add(root.left);
                if (root.right != null) q.add(root.right);
            }
        }

        return ret;
    }
}
